### PR TITLE
Use traceError instead of error for better error messages (fix #3608).

### LINKED ISSUE #3608: `Builtins.error` should accepts a string as input
## Area

[x] *Plutus Foundation* Related to the GHC plugin, Haskell-to-Plutus compiler, on-chain code
[] *Plutus Application Framework* Related to the Plutus application backend (PAB), emulator, Plutus libraries
[] *Marlowe* Related to Marlowe
[] *Other* Any other topic (Playgrounds, etc.)

## Describe the feature you'd like

Currently, many function in Plutus throw `Builtins.error ()` with no description which is very difficult to debug. We get cryptic errors such as: `WalletError (ValidationError (ScriptFailure (EvaluationError [])))` in phase 2 validation.
I suggest that `Builtins.error` should accept a description string as a parameter.
https://github.com/input-output-hk/plutus/blob/master/plutus-tx/src/PlutusTx/Builtins.hs#L190

### PR DESCRIPTION
**Pull request by ak3n**
This PR replaces `error ()` with `traceError` providing error messages for better debugging. (fix #3608).

---

Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

Pre-merge checklist:
- [x] Someone approved it
- [x] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

**Review by michaelpj — APPROVED** [review]
May need to update the output of some tests too.

File: plutus-tx/src/PlutusTx/List.hs [hunk 1]
```diff
@@ -101,7 +103,8 @@ findIndex p l = listToMaybe (findIndices p l)
 --
 infixl 9 !!
 (!!) :: [a] -> Integer -> a
-[]       !! _ = Builtins.error ()
+_        !! n | n < 0 = traceError "PlutusTx.List.!!: negative index"
```
Comment (michaelpj) [inline plutus-tx/src/PlutusTx/List.hs:106]: This is using the Haskell prelude `<` instead of the Plutus Tx one.

Comment (ak3n) [reply plutus-tx/src/PlutusTx/List.hs:106]: Oh, good catch! Thanks!

Comment (ak3n) [reply plutus-tx/src/PlutusTx/List.hs:107]: It forced to move `Min(..)` and `Max(..)` to `PlutusTx.Semigroup`.

Comment (michaelpj) [reply plutus-tx/src/PlutusTx/List.hs:107]: Fine. Alternatively you could use the builtin directly, but either way is good.

**Comment by michaelpj** [comment]
Yeah, you're going to need to update the `plutus-use-cases` test output. Sorry, this will take ages since it has to build the node :(
```
cabal test plutus-use-cases --test-options=--accept
```

**Pull request merged**